Fix/syndic/62618 - external auth

waynew opened the pull request:
### What does this PR do?

Fixes #62618 - syndic eauth 

### Previous Behavior

Syndics didn't work with eauth... now they do. It also fixes #62577. And adds a fairly comprehensive integration test suite for syndics.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

---

I'm definitely going to be refactoring chunks of this, but I wanted to get this PR open in case anyone had comments for stuff that either I've missed or could do better.

Bits I still have planned:

- [x] - remove/redo logging
- [x] - consider how I'm firing the event in _list_valid_minions - is there a better way to get that from the syndic master to the salt-syndic service?
- [x] - refactor the do_something and order_masters bits of code around that.
- [x] - restore some of the authorization error messages
- [x] - adjust the path for communication if I can't figure out a way to get it on the event bus
- [x] - take a closer look at the AsyncReqChannel that enables returns. It may or may not be the best way to do that.

waynew commented on salt/channel/client.py:
```diff
@@ -556,6 +556,7 @@ def _decode_payload(self, payload):
             try:
                 payload["load"] = self.auth.crypticle.loads(payload["load"])
             except salt.crypt.AuthenticationError:
+                log.warn('Failed authing this payload: %r', payload)
```
TODO: remove this

waynew commented on salt/client/__init__.py:
```diff
@@ -348,6 +348,7 @@ def _check_pub_data(self, pub_data, listen=True):
             return pub_data
 
         if self.opts.get("order_masters"):
+            # print('Listening for ', pub_data['jid'], datetime.now())
```
TODO: will remove this

waynew commented on salt/client/__init__.py:
```diff
@@ -1137,6 +1140,13 @@ def get_iter_returns(
                 minions = {minions}
             elif isinstance(minions, (list, tuple)):
                 minions = set(list(minions))
+        expected_syndics = set()
+        if self.opts["order_masters"]:
+            if not isinstance(kwargs["syndics"], set):
+                if isinstance(kwargs["syndics"], str):
+                    expected_syndics = {kwargs["syndics"]}
+                elif isinstance(kwargs["syndics"], (list, tuple)):
+                    expected_syndics = set(list(kwargs["syndics"]))
```
**question**: Is this the best way to get the behavior here in the command line client? I *think* so. Though there may also still be some other legacy syndic specific code in here...

waynew commented on salt/client/__init__.py:
```diff
@@ -1986,13 +1993,98 @@ def pub_async(
+                self.opts["sock_dir"],
+            )
+            raise SaltClientError
+        payload_kwargs, payload = payload, None
```
I'm probably going to reorder these if it turns out that I have to have them.

I *think* I do, because `pub_async` doesn't actually offer the ability to publish more or less arbitrary data. And maybe this is the wrong way to communicate! So if anyone has some better ideas, I'm all ears.

waynew commented on salt/client/__init__.py:
```diff
@@ -2002,6 +2094,7 @@ def pub_async(
                 # If not, we won't get a response, so error out
                 if listen and not self.event.connect_pub(timeout=timeout):
                     raise SaltReqTimeoutError()
+                log.warn("SENDING %r", payload_kwargs)
```
Yep, I'll be removing this logging code and the one above (or providing some better logging :thinking: )

waynew replied:
gone

waynew commented on salt/crypt.py:
```diff
@@ -813,7 +813,7 @@ def handle_signin_response(self, sign_in_payload, payload):
             master_pubkey_path, clear_signed_data, clear_signature
         ):
-            log.critical("The payload signature did not validate.")
+            log.critical("The payload signature did not validate. XOXO %r", clear_signed_data)
```
Will restore this one

waynew replied:
restored

waynew commented on salt/minion.py:
```diff
@@ -1620,12 +1633,31 @@ def _send_req_async(self, load, timeout):
+            ret = yield channel.send(
+                load, timeout=timeout, tries=self.opts["return_retry_tries"]
+            )
+
```
For some reason, *this* is what appears to make #62933 work. But it's not great, because the syndic is going to have to create a new connection every time a req is sent to the MoM? :thinking:

waynew replied:
Ahhhh... never mind I guess that should actually be the *local* Syndic Master.

waynew replied:
:+1: Done

waynew commented on salt/minion.py:
```diff
@@ -3289,7 +3338,7 @@ def syndic_cmd(self, data):
 
         def timeout_handler(*args):
-            log.warning("Unable to forward pub data: %s", args[1])
+            log.warning("Unable to forward pub data: %s %r", args[1], data)
```
will be restoring this

waynew replied:
restored

waynew commented on salt/minion.py:
```diff
@@ -3641,6 +3691,8 @@ def _process_event(self, raw):
         mtag, data = self.local.event.unpack(raw)
         log.trace("Got event %s", mtag)  # pylint: disable=no-member
+        if 'tgt_return' in mtag:
+            data["syndic_id"] = self.opts["id"]
```
I'm not sure if there's a better way to get the ID of the syndic that's returning this data to the MoM.

waynew replied:
After playing around a little, pretty sure the answer is "no". The only place that the syndic/minion id will be present is in the syndic service itself. Without publishing an event per syndic (ew, gross), basically the only way is to inject the data into the event stream when we go to forward it onto the MoM.

I *think* I'm actually seeing the event from the master as *well* as the forwarded syndic event show up on the MoM, which is uh... weird. But it doesn't seem to have a huge impact.

waynew commented on salt/modules/state.py:
```diff
@@ -2490,12 +2490,18 @@ def event(
+        print(sevent.subscriber.stream.fileno())
+        import os
+        print(sevent.subscriber)
+        print(os.stat(sevent.subscriber.socket_path).st_ino)
```
debugging code, to be removed

waynew replied:
gone

waynew commented on salt/returners/local_cache.py:
```diff
@@ -281,7 +283,9 @@ def get_load(jid):
     load_fn = os.path.join(jid_dir, LOAD_P)
+    # log.warn("SAVING LOAD Looking for %r", load_fn)
     if not os.path.exists(jid_dir) or not os.path.exists(load_fn):
+        # log.warn("LOL NOPE %r %r", os.path.exists(jid_dir), os.path.exists(load_fn))
```
log calls will be removed

waynew replied:
gone

waynew commented on salt/transport/ipc.py:
```diff
@@ -733,6 +736,7 @@ def read(self, timeout):
         :param callback: A callback with the received data
         """
         if self._saved_data:
+            #log.warn('YEP! I CAN HAZ SAVED %r %r', id(self), id(self.read))
```
also log calls to be removed

waynew replied:
gone

waynew commented on salt/utils/minions.py:
```diff
@@ -691,6 +691,13 @@ def _all_minions(self, expr=None):
+        syndic_cache_path = os.path.join(
+            self.opts["cachedir"], "syndics"
+        )
+        return [f for f in os.listdir(syndic_cache_path)]
```
I need to get a list of syndics that are attached to this particular master. If there's a better way here, I'm all for it!

waynew replied:
Note: this is somewhat unreliable, in that the syndics aren't present until there's a `_syndic_return`

But it can easily be primed with a `salt \* test.ping` on the MoM

waynew replied:
Adding the timeout functionality and syndic job checking actually turns out to make the priming unnecessary. It's still probably a good idea to check this, but we also end out publishing the job and aggregating all of the syndics that return.

waynew commented on salt/utils/minions.py:
```diff
@@ -973,6 +980,65 @@ def auth_check(
         return False
 
+
+    def check_syndic_auth_list(self, *, minions, funs, args, auth_list):
```
this is basically auth_check, but with less pieces. In particular, IIRC, auth_check ends out looking at the PKI dir. We can't do that - we have to compare the minions returned by syndics to their eauth/acl.

So this pulls out all of the code that deals with that, and just compares the functions and commands.

waynew committed:
add integration tests for syndic eauth

waynew committed:
Convert syndics to use older AsyncReqChannel

There *may* be a better way to get results from syndics to the Master of
Masters, but this at least works.

Fixes #62933

waynew committed:
refactor publish, extract publisher_acl

Aiming to simplify publish where possible, to make it easier to follow.

waynew committed:
Fix Syndic eauth failure

This is a complete restructure of how Master-of-Masters and Syndics
communicate when publishing a job.

In the past, if we were a MoM, the undocumented but desired approach was
that the MoM would not have any minions attached to it. Because when a
Salt Master was a MoM it would simply publish the job 100% of the time
unless the ACL denied the command. But this was not based on whether or
not there were minions that could be targeted. In fact, when no valid
minions were found then the job would be published, it was only when
there were *invalid* minions found that things would be rejected.

This new approach is for the MoM to make a request of the syndics to
return a list of minions that match both the allowed ACL targets as well
as the requested targets.

On the MoM if the user has requested more minions than they have access
to then we will fail on authorization.

There are a few known issues with the existing architecture:

1. It's possible (though undocumented as to what the behavior should be)
   for multiple syndics to have different minions with the same ID.
2. Based on timing, and without a complete rearchitecture of Syndics
   (planned for 2023), if a syndic comes online *after* targeting data
   is returned and *before* the job is published it's possible that a
   user may be able to publish jobs to minions they should not have
   access to.
3. It's undocumented how things should work if a Syndic is a MoM.

But now - when using external auth, or publisher_acl, the only time a
job should be published is when the provided user actually has the
correct ACL to publish the jobs.

waynew commented on salt/exceptions.py:
```diff
@@ -344,6 +344,13 @@ class EauthAuthenticationError(SaltException):
+    to perform the requested action.
+    """
+
+
```
This was added when I realized that Authentication != Authorization. The previous failure is failure to login, at this point (generally speaking) the external auth user has been authenticated, they're just not **Authorized** to perform an action, and thus :point_up: this exception.

waynew commented on salt/master.py:
```diff
@@ -1950,6 +1950,7 @@ class ClearFuncs(TransportMethods):
         "mk_token",
         "wheel",
         "runner",
+        "_list_valid_minions",
```
**question** does `_list_valid_minons` belong here in ClearFuncs? My *thought* is that it's fine here since it's not much different than `ping`, and is used from `publish`. But I would happily be educated as to why that thought is incorrect.

waynew commented on salt/master.py:
```diff
@@ -1979,6 +1980,40 @@ def __init__(self, opts, key):
+        # self.event.fire_event(
+        #    {"Return Some Things": jid, 'valid_minions': list(valid_minions), 'target_minions': tgt_minions},
+        #    tagify([jid, 'tgt_return'], "syndic"),
+        # )
```
**chore** probably should remove this comment. I'm still not entirely clear when events need "fun" and "ret". I *think* I could skip a `_prep_pub`. Also... 

**question** does prepping a jid here do anything silly? :thinking:

waynew replied:
done. No longer prepping jid, so not actually relevant now.

waynew commented on salt/client/__init__.py:
```diff
@@ -1935,7 +1931,83 @@ def pub(
+    ):
+        """
+        Publish an arbitrary payload. Maybe this is what events do?
+        Publish an arbitrary payload. Maybe this is what events do?
```
**suggestion** somehow I managed to duplicate a doc line here. Fix this docstring.

waynew commented on salt/minion.py:
```diff
@@ -3378,6 +3393,39 @@ def destroy(self):
+            ret = yield channel.send(
+                load, timeout=timeout, tries=self.opts["return_retry_tries"]
+            )
+            raise salt.ext.tornado.gen.Return(ret)
```
**note** this is mostly just copied verbatim from minion, but the channel was changed to be able to return reqs to MoM rather than the Syndic master.

waynew commented on salt/returners/local_cache.py:
```diff
@@ -297,7 +298,8 @@ def get_load(jid):
         log.critical("Failed to unpack %s", load_p)
-        raise exc
+        if exc is not None:
+            raise exc
```
**note** this was an unrelated bug that was exposed during work here.

waynew commented on salt/master.py:
```diff
@@ -1979,6 +1980,40 @@ def __init__(self, opts, key):
+        jid = self._prep_jid(event_ret, {})
+        payload = self._prep_pub([], jid, event_ret, {}, missing)
+        self.event.fire_event(payload, tagify([jid, "tgt_return"]))
+        # self.event.fire_event(
```
See my other comment - one or the other :joy: 

I haven't gone back and tested, but now that my tests are all passing, and passing in ~5m it's going to be a *lot* easier to test, but do you know if it matters if I just fire an event vs create the payload with prepping the jid, etc?

My gut says that it would actually be better to just fire an event rather than do all the job bits... but again I haven't tested it yet. I have a vague memory that at one point I tried it and I could see the event over on the `_collect_syndics` event watcher.

waynew replied:
Okay, so experiments verify that I can, in fact, avoid doing the prep_pub and prep_jid steps and simply fire off an event! Seems to simplify things quite a bit, too.

waynew commented on salt/master.py:
```diff
@@ -2268,10 +2347,92 @@ def publish(self, clear_load):
+            ret["syndics"] = syndics
+        return ret
+
+    def _collect_syndics(self, *, clear_load, extra, minions, auth_list, auth_type):
```
That's a good question. I think that there's some mechanism that will time this out higher up the stack. One interesting phenomena I've seen is that the CLI may report that the job didn't really run, even though events actually get published and the minions return back to the syndics which return back to the MoM in a sort of eventual consistency. At the time I *was* timing out on the `while` loop -- which behavior I *could* restore. Though I'm not entirely sure the best timeout to add for that. The `clear_load["to"]` contains a timeout value in seconds, but *other* places up the stack also use that value so... I could use it as well, or use half or some fraction of it? :shrug: My current experience doesn't give me any strong opinions on it.

waynew replied:
I went ahead and added the timeout logic in there. Should still work the same, and roughly the same as it used to (that is, if the syndic was offline you just kinda get a :shrug: from salt and it may or may not be thrilled about it)

waynew commented on salt/client/__init__.py:
```diff
@@ -1414,13 +1410,13 @@ def get_returns(self, jid, minions, timeout=None):
             if len(found.intersection(minions)) >= len(minions):
                 # All minions have returned, break out of the loop
-                log.debug("jid %s found all minions", jid)
+                log.debug("jid %s found all minions")
```
Oops. I had some debugging code in, and apparently was too aggressive with removing it :sweat_smile:

waynew commented on salt/client/__init__.py:
```diff
@@ -1414,13 +1410,13 @@ def get_returns(self, jid, minions, timeout=None):
                 if len(found.intersection(minions)) >= len(minions):
                     # All minions have returned, break out of the loop
-                    log.debug("jid %s found all minions", jid)
+                    log.debug("jid %s found all minions")
```
Same as other comment, will restore

waynew committed:
Tests allowed on Python <3.10

waynew committed:
Update Syndic docs

Yeah, maybe Syndics have been around for a while but a lot of their
behavior is not well-defined.

waynew committed:
Add changelog entries

waynew commented:
@Ch3LL changelog files have been added. I think I may have pushed them before but I was trying to rebase and I lost them in the shuffle :sweat: 

And I haven't completely rebased either :joy: :sweat_smile:

waynew committed:
SQUASHME

waynew committed:
Remove experimental note

Required by Anil

waynew commented on doc/topics/topology/syndic.rst:
```diff
@@ -4,31 +4,46 @@
 Salt Syndic
 ===========
 
+.. warning::
+
+    Salt Syndic with :ref:`external auth or publisher_acl<acl-eauth>`
+    should not be used for security purposes.
+
+    The long-term goal for Syndics is that they work correctly with
+    authentication and authorization, but with Salt's current architecture that
+    Syndics work with publisher_acl/external auth at all should be considered a
+    convenience. Syndics with external auth or publisher_acl should not used
+    anywhere that security is considered essential. With the changes resulting
+    from `issue #62618 on GitHub
+    <https://github.com/saltstack/salt/issues/62618>`_, there have been some
+    improvements, but currently Syndic functionality is only fully supported
+    when running ``salt`` on the Master of Masters as the root/service user.
+
+    Additionally, it's technically possible to have minions of the same
+    name on different Syndics. In that case, Salt's behavior is completely
+    undefined and subject to change.
+
+    It may also be possible for a Syndic to itself be a Master of Masters, but
+    that behavior is also undocumented and should not yet be considered a
+    supported feature.
```
@naaronne Moving & updating the below warning - Anil wanted to get your eyes on this since we're explicitly calling out that Syndics, despite their inclusion in Salt for several years, should not (yet) be treated as a security feature.

Before, we simply lacked any documentation about whether or not, or how Syndics might interact with publisher_acl or eauth.

This PR *does* restore the previous functionality, re-enabling eauth, but until we can take the time to really refactor Syndics, I would like to warn against relying on it for security.

anilsil replied:
This documentation needs to be rewritten, it doesn't reflect the confidence of this fix.

waynew replied:
FWIW, this documentation is not documenting the fix as much as the general state of Syndics.

An analogy I thought of is that Syndics could be considered a chainsaw. There are plenty of people who use chainsaws on a daily basis, to no ill-effects, because they understand the dangers, wear the proper protective equipment, and have a safety attitude (e.g. don't operate while mentally/chemically impaired).

However, if you're not wearing the correct equipment, and you operate a chainsaw dangerously, you might take your leg off.

Just because I replaced the broken chain, doesn't mean that the chainsaw ceases to be dangerous. And putting up a new warning sign about the dangers of the chainsaw doesn't change anything about the chainsaw being a chainsaw. And it shouldn't imply that we're not confident about the chain. 

But I wouldn't go around juggling the chainsaw :slightly_smiling_face:

waynew committed:
Skip if docker not found

Do our tests care we cannot connect to docker in the environment? We do
not. If docker can't get the client to connect it's OK, we'll just skip
the tests. At least *one* of our platforms should be running these
tests, which is enough.

waynew committed:
Also skip when keys cannot be accepted

waynew committed:
Rewrite warning

waynew committed:
Another test skip

No need to fail when we can't clean up. VMs *should* be going away, and
if you're running locally you can clean up yourself :+1:

twangboy commented on doc/topics/topology/syndic.rst:
```diff
@@ -4,31 +4,37 @@
+A Syndic node is a special passthrough Minion node.  A Syndic node consists of
+a ``salt-syndic`` daemon and a ``salt-master`` daemon running on the same
+system.  The ``salt-master`` daemon running on the Syndic node controls a group
+of lower level Minion nodes and the ``salt-syndic`` daemon connects higher
```
the ``salt-syndic`` daemon connects to a higher

waynew replied:
Gotta love long lived typos. Fixed :+1:

twangboy reviewed (changes requested):
Just a minor docs change

waynew committed:
Fix existing typo

waynew commented:
Looks like at least one of these failures is https://github.com/saltstack/salt/commit/9f95af801e2fb0b12426722037f41a2ce2205576

waynew committed:
Merge branch 'master' into fix/syndic/62618

waynew committed:
Merge branch 'master' into fix/syndic/62618

garethgreenaway committed:
Merge branch 'master' into fix/syndic/62618

waynew committed:
Merge branch 'master' into fix/syndic/62618

waynew committed:
Merge branch 'master' into fix/syndic/62618

waynew committed:
bump up the timeout

Ch3LL commented:
@waynew can we close this one in favor of https://github.com/saltstack/salt/pull/63428 ?

waynew commented:
Yup - #63428 deprecates? obviates? the need for this PR.
